js/src/index.ts

```diff
@@ -392,28 +357,66 @@ export class Schematic {
       };
 
       if (socket.readyState === WebSocket.OPEN) {
+        // If websocket is already open, send message immediately
         sendMessage();
       } else if (socket.readyState === WebSocket.CONNECTING) {
+        // If websocket is connecting, wait for it to open before sending message
         socket.addEventListener("open", sendMessage);
       } else {
+        // If websocket is closed, reject the promise
         reject("WebSocket is not open or connecting");
       }
     });
   };
-}
 
-function contextString(context: SchematicContext): string {
-  const sortedContext = Object.keys(context).reduce((acc, key) => {
-    const sortedKeys = Object.keys(
-      context[key as keyof SchematicContext] || {},
-    ).sort();
-    const sortedObj = sortedKeys.reduce((obj, sortedKey) => {
-      obj[sortedKey] = context[key as keyof SchematicContext]![sortedKey];
-      return obj;
-    }, {} as Keys);
-    acc[key as keyof SchematicContext] = sortedObj;
-    return acc;
-  }, {} as SchematicContext);
-
-  return JSON.stringify(sortedContext);
+  /**
+   * State management
+   */
+
+  // isPending state
+  getIsPending = (): boolean => {
+    return this.isPending;
+  };
+
+  addIsPendingListener = (listener: () => void) => {
+    this.isPendingListeners.add(listener);
+    return () => {
+      this.isPendingListeners.delete(listener);
+    };
+  };
+
+  private setIsPending = (isPending: boolean) => {
+    this.isPending = isPending;
+    this.isPendingListeners.forEach((listener) => listener());
+  };
+
+  // flagValues state
+  getFlagValue = (flagKey: string) => {
+    const values = this.getFlagValues();
+    return values[flagKey] ?? false;
+  };
+
+  getFlagValues = () => {
+    const contextStr = contextString(this.context);
+    return this.values[contextStr] ?? {};
```

Comment:
Remove `?? {}` - we want to return undefined in this case so that callers can use fallback values. schematic-react implements fallback values in useSchematicFlag